---
title: "**S-LDSC and Torus analysis to estimate the enrichment of GWAS variants in m6A peaks**"
author: "Christie Du"
date: "2020-07-30"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## **Introduction**

I ran S-LDSC and Torus on m6A annotations of 8 tissues: brain, heart, kidney, liver, lung, muscle, placenta, and stomach. I did this for 11 GWAS: Asthma, BMI, CAD, HDL, Height, Intelligence, IBD, LDL, RA, Schizophrenia, and Type II Diabetes. I also ran Torus for blood pressure GWAS: diastolic blood pressure (DBP), pulse pressure (DBP), and systolic blood pressure (SBP). 

m6A annotations used are from this paper: Xiao et al. The RNA N6-methyladenosine modification landscape of human fetal tissues. Nature 2019.

## **S-LDSC**

For S-LDSC, I used two baseline models: 

1. A subset of the baseline LD annotations located here: `RCC:/project2/xinhe/kevinluo/ldsc/annot/ldscores/baseline_gene_MAF_LD/`
2. Baseline LD v1.1 located here:
`RCC:/project2/xinhe/kevinluo/ldsc/LDSCORE/1000G_Phase3_baselineLD_v1.1_ldscores`

### GWAS and m6A annotations:
* GWAS summary statistics in the sumstats format (excluding Intelligence) are located at `RCC:/project2/xinhe/kevinluo/GWAS/GWAS_summary_stats/GWAS_collection/ldsc_format` 
* Intelligence GWAS in the sumstats format is located at `RCC:/project2/xinhe/kevinluo/GWAS/GWAS_summary_stats/GWAS_from_Min/ldsc_format` 
* The original m6A annotations in BED format are located at `RCC:/project2/xinhe/cdu/ldsc/results/sLDSC_Xiao_m6A_2019/Annot/hg38/Bed` 

Because the GWAS summary statistics were based on the hg19 map, the m6A annotations were converted from hg38 to hg19 using `liftOver`. 

```{r annotations and traits}
annot_list <- c("brain", "heart", "kidney", "liver", "lung", "muscle", "placenta", "stomach")
trait_name_list <- c("Asthma", "BMI","CAD","HDL","Height","IBD","Intelligence","LDL","RA","Schizophrenia","T2D")
```


```{r plot functions}
library(ggplot2)
library(foreach)
library(doParallel)

ggplot_enrichment <- function(result_sLDSC, xlim = NULL, title = "Enrichment"){
  
  Enrichment <- result_sLDSC$Enrichment
  Enrichment_CI_L <- result_sLDSC$Enrichment - 1.96*result_sLDSC$Enrichment_std_error
  Enrichment_CI_H <- result_sLDSC$Enrichment + 1.96*result_sLDSC$Enrichment_std_error

  ## truncate at 1
  Enrichment[Enrichment < 1] <- 1
  Enrichment_CI_L[Enrichment_CI_L < 1] <- 1
  Enrichment_CI_H[Enrichment_CI_H < 1] <- 1
  
  p <- ggplot(result_sLDSC, aes(x = Enrichment, y = Category, colour = Color))+
    geom_point()+
    xlab("Enrichment")+
    ggtitle(title)+
    geom_errorbarh(aes(xmin = Enrichment - 1.96*Enrichment_std_error, 
                       xmax = Enrichment + 1.96*Enrichment_std_error, height = 0.1))+ 
    facet_wrap(Disease~.,ncol = 4)+
    theme_bw()  + 
    geom_vline(xintercept = 1,linetype="dotted", colour = "red")+
    theme(axis.ticks = element_blank(),  
          panel.grid.minor = element_blank(), 
          axis.line = element_line(colour = "black"), 
          axis.text = element_text(face="bold",size = 9, colour = "black"),
          axis.title = element_text(face="bold",size = 9),
          strip.text = element_text(face="bold",size = 9), 
          panel.spacing.x = unit(0.6,units = "cm"), 
          axis.title.y = element_blank(), 
          legend.position = "none", 
          plot.title = element_text(hjust = 0.5))
  if(!is.null(xlim)){
    p <- p + coord_cartesian(xlim = xlim)
  }
  print(p)
}

ggplot_log2_enrichment <- function(result_sLDSC, xlim = NULL, title = "Enrichment"){
  
  result_sLDSC$Enrichment_CI_L <- result_sLDSC$Enrichment - 1.96*result_sLDSC$Enrichment_std_error
  result_sLDSC$Enrichment_CI_H <- result_sLDSC$Enrichment + 1.96*result_sLDSC$Enrichment_std_error

  ## truncate at 1
  result_sLDSC$Enrichment[result_sLDSC$Enrichment < 1] <- 1
  result_sLDSC$Enrichment_CI_L[result_sLDSC$Enrichment_CI_L < 1] <- 1
  result_sLDSC$Enrichment_CI_H[result_sLDSC$Enrichment_CI_H < 1] <- 1

  p <- ggplot(result_sLDSC, aes(x = log2(Enrichment), y = Category, colour = Color))+
    geom_point()+
    xlab("log2(Enrichment)")+
    ggtitle(title)+
    geom_errorbarh(aes(xmin = log2(Enrichment_CI_L), 
                       xmax = log2(Enrichment_CI_H), height = 0.1))+ 
    facet_wrap(Disease~.,ncol = 4)+
    theme_bw()  + 
    geom_vline(xintercept = 0,linetype="dotted", colour = "red")+
    theme(axis.ticks = element_blank(),  
          panel.grid.minor = element_blank(), 
          axis.line = element_line(colour = "black"), 
          axis.text = element_text(face="bold",size = 9, colour = "black"),
          axis.title = element_text(face="bold",size = 9),
          strip.text = element_text(face="bold",size = 9), 
          panel.spacing.x = unit(0.6,units = "cm"), 
          axis.title.y = element_blank(), 
          legend.position = "none", 
          plot.title = element_text(hjust = 0.5))
  if(!is.null(xlim)){
    p <- p + coord_cartesian(xlim = xlim)
  }
  print(p)
}

ggplot_heritability <- function(result_sLDSC, xlim = NULL, title = "Heritability"){
  ## Proportion of heritability
  p <- ggplot(result_sLDSC, aes(x = Prop._h2*100, y = Category, colour = Color))+
    geom_point()+
    xlab("Heritability %")+
    ggtitle(title)+
    geom_errorbarh(aes(xmin = (Prop._h2-1.96*Prop._h2_std_error)*100, 
                       xmax = (Prop._h2+1.96*Prop._h2_std_error)*100, height = 0.1))+ 
    facet_wrap(Disease~.,ncol = 4)+
    theme_bw()  + 
    geom_vline(xintercept = 0,linetype="dotted", colour = "red")+
    theme(axis.ticks = element_blank(),  
          panel.grid.minor = element_blank(), 
          axis.line = element_line(colour = "black"), 
          axis.text = element_text(face="bold",size = 12, colour = "black"),
          axis.title = element_text(face="bold",size = 12),
          strip.text = element_text(face="bold",size = 12), 
          panel.spacing.x = unit(0.6,units = "cm"), 
          axis.title.y = element_blank(), 
          legend.position = "none", 
          plot.title = element_text(hjust = 0.5))
  if(!is.null(xlim)){
    p <- p + coord_cartesian(xlim = xlim)
  }
  print(p)
}

## combine S-LDSC enrichment results across traits
combine_sldsc_traits <- function(trait_name_list, dir_results, baseline){
  registerDoParallel(cores = 10)
  
  result_sLDSC <- foreach(trait = trait_name_list, .combine = rbind)%dopar%{
    sldsc_results <- read.table(paste0(dir_results,"/", trait, "/", baseline, "/", trait,"_", "brain", "_", baseline, ".results"), header = T, stringsAsFactors = F)
    sldsc.brain <- sldsc_results[sldsc_results$Category == "L2_1",]
    sldsc.brain$Category <- "brain"
    
    sldsc_results <- read.table(paste0(dir_results,"/", trait, "/", baseline, "/", trait,"_", "heart", "_", baseline, ".results"), header = T, stringsAsFactors = F)
    sldsc.heart <- sldsc_results[sldsc_results$Category == "L2_1",]
    sldsc.heart$Category <- "heart"
    
    sldsc_results <- read.table(paste0(dir_results,"/", trait, "/", baseline, "/", trait,"_", "kidney", "_", baseline, ".results"), header = T, stringsAsFactors = F)
    sldsc.kidney <- sldsc_results[sldsc_results$Category == "L2_1",]
    sldsc.kidney$Category <- "kidney"
    
    sldsc_results <- read.table(paste0(dir_results,"/", trait, "/", baseline, "/", trait,"_", "liver", "_", baseline, ".results"), header = T, stringsAsFactors = F)
    sldsc.liver <- sldsc_results[sldsc_results$Category == "L2_1",]
    sldsc.liver$Category <- "liver"
    
    sldsc_results <- read.table(paste0(dir_results,"/", trait, "/", baseline, "/", trait,"_", "lung", "_", baseline, ".results"), header = T, stringsAsFactors = F)
    sldsc.lung <- sldsc_results[sldsc_results$Category == "L2_1",]
    sldsc.lung$Category <- "lung"
    
    sldsc_results <- read.table(paste0(dir_results,"/", trait, "/", baseline, "/", trait,"_", "muscle", "_", baseline, ".results"), header = T, stringsAsFactors = F)
    sldsc.muscle <- sldsc_results[sldsc_results$Category == "L2_1",]
    sldsc.muscle$Category <- "muscle"
    
    sldsc_results <- read.table(paste0(dir_results,"/", trait, "/", baseline, "/", trait,"_", "placenta", "_", baseline, ".results"), header = T, stringsAsFactors = F)
    sldsc.placenta <- sldsc_results[sldsc_results$Category == "L2_1",]
    sldsc.placenta$Category <- "placenta"
    
    sldsc_results <- read.table(paste0(dir_results,"/", trait, "/", baseline, "/", trait,"_", "stomach", "_", baseline, ".results"), header = T, stringsAsFactors = F)
    sldsc.stomach <- sldsc_results[sldsc_results$Category == "L2_1",]
    sldsc.stomach$Category <- "stomach"
    
    sldsc.combined <- rbind(sldsc.brain, sldsc.heart, sldsc.kidney, sldsc.liver, sldsc.lung, sldsc.muscle, sldsc.placenta, sldsc.stomach)
    sldsc.combined <- cbind(Disease = trait, sldsc.combined)
    sldsc.combined
  }
  return(result_sLDSC)
}
```

### Modified baseline model

```{r data import}
baseline <- "baseline_gene_MAF_LD"

dir_results <- "/project2/xinhe/cdu/ldsc/results/sLDSC_Xiao_m6A_2019/results_sLDSC_hg19"
result_sLDSC <- combine_sldsc_traits(trait_name_list, dir_results, baseline)

result_sLDSC$Category <- factor(result_sLDSC$Category, levels = rev(annot_list) )
result_sLDSC$Color <- factor(result_sLDSC$Category, levels = annot_list)
```


```{r data table}
DT::datatable(format(result_sLDSC[,1:8], digits = 2), options = list(scrollX = TRUE, keys = TRUE, pageLength = length(annot_list)),rownames = F)
```


```{r plots}
ggplot_enrichment(result_sLDSC, title = "")
ggplot_log2_enrichment(result_sLDSC, title = "")
```

### Baseline V1.1

```{r v1.1 data import}
baseline <- "baselineLDv1.1"

dir_results <- "/project2/xinhe/cdu/ldsc/results/sLDSC_Xiao_m6A_2019/results_sLDSC_hg19"
result_sLDSC <- combine_sldsc_traits(trait_name_list, dir_results, baseline)

result_sLDSC$Category <- factor(result_sLDSC$Category, levels = rev(annot_list) )
result_sLDSC$Color <- factor(result_sLDSC$Category, levels = annot_list)
```


```{r v1.1 data table}
DT::datatable(format(result_sLDSC[,1:8], digits = 2), options = list(scrollX = TRUE, keys = TRUE, pageLength = length(annot_list)),rownames = F)
```


```{r v1.1 plots}
ggplot_enrichment(result_sLDSC, title = "")
ggplot_log2_enrichment(result_sLDSC, title = "")
```

## **Torus**

### GWAS and m6A annotations:
* Asthma, BMI, HDL, Height, LDL, RA, and Schizophrenia GWAS are located at `RCC:/project2/xinhe/kevinluo/GWAS/GWAS_summary_stats/GWAS_collection/GWAS_raw` 
* The Intelligence GWAS is located at `RCC:/project2/xinhe/kevinluo/GWAS/GWAS_summary_stats/GWAS_from_Min/raw_data` 
* CAD, blood pressure, and IBD GWAS are located at `RCC:/project2/compbio/gwas_summary_statistics`
* The liftOver hg19 m6A annotations are located at `RCC:/project2/xinhe/cdu/torus/annot/Xiao/hg19`
* GWAS data suitable for Torus can be found here `RCC:/project2/xinhe/cdu/GWAS/Xiao_torus_format`

There are two BMI GWAS datasets: `BMI_GWAS_fromScott.txt.gz` and `BMI_GWAS_raw`. The `BMI_GWAS_raw` dataset was missing the chromosome and position data, so the missing SNP positions were taken from `BMI_GWAS_fromScott.txt.gz`. Of the 2554637 SNPs reported in `BMI_GWAS_raw`, 2092511 matched those in `BMI_GWAS_fromScott.txt.gz`.

Running Torus on T2D GWAS failed because too few T2D variants were matched - only 10,428 out of 2,470,851 variants were matched.

```{r torus annotations and traits}
annot_list <- c("brain", "heart", "kidney", "liver", "lung", "muscle", "placenta", "stomach")
trait_name_list <- c("ASTHMA","BMI","CAD","HDL","HEIGHT","IBD","INTELLIGENCE","LDL","RA","SCZ","DBP","PP","SBP")
```


```{r torus functions}
library(ggplot2)
library(foreach)
library(doParallel)

ggplot_enrichment <- function(result_torus, xlim = NULL, title = "Enrichment"){
  
  Enrichment <- result_torus$Enrichment
  Enrichment_CI_L <- result_torus$Low
  Enrichment_CI_H <- result_torus$High
  
  p <- ggplot(result_torus, aes(x = Enrichment, y = Category, colour = Color))+
    geom_point()+
    xlab("Enrichment")+
    ggtitle(title)+
    geom_errorbarh(aes(xmin = Enrichment_CI_L, 
                       xmax = Enrichment_CI_H, height = 0.1))+ 
    facet_wrap(Disease~.,ncol = 4)+
    theme_bw()  + 
    geom_vline(xintercept = 1,linetype="dotted", colour = "red")+
    theme(axis.ticks = element_blank(),  
          panel.grid.minor = element_blank(), 
          axis.line = element_line(colour = "black"), 
          axis.text = element_text(face="bold",size = 9, colour = "black"),
          axis.title = element_text(face="bold",size = 9),
          strip.text = element_text(face="bold",size = 9), 
          panel.spacing.x = unit(0.6,units = "cm"), 
          axis.title.y = element_blank(), 
          legend.position = "none", 
          plot.title = element_text(hjust = 0.5))
  if(!is.null(xlim)){
    p <- p + coord_cartesian(xlim = xlim)
  }
  print(p)
}

ggplot_log2_enrichment <- function(result_torus, xlim = NULL, title = "Enrichment"){
  
  result_torus$Enrichment_CI_L <- log2(result_torus$Low)
  result_torus$Enrichment_CI_H <- log2(result_torus$High)

  ## truncate at 1
  # result_torus$Enrichment[result_torus$Enrichment < 1] <- 1
  # result_torus$Enrichment_CI_L[result_torus$Enrichment_CI_L < 1] <- 1
  # result_torus$Enrichment_CI_H[result_torus$Enrichment_CI_H < 1] <- 1

  p <- ggplot(result_torus, aes(x = log2(Enrichment), y = Category, colour = Color))+
    geom_point()+
    xlab("log2(Enrichment)")+
    ggtitle(title)+
    geom_errorbarh(aes(xmin = Enrichment_CI_L, 
                       xmax = Enrichment_CI_H, height = 0.1))+ 
    facet_wrap(Disease~.,ncol = 4)+
    theme_bw()  + 
    geom_vline(xintercept = 0,linetype="dotted", colour = "red")+
    theme(axis.ticks = element_blank(),  
          panel.grid.minor = element_blank(), 
          axis.line = element_line(colour = "black"), 
          axis.text = element_text(face="bold",size = 9, colour = "black"),
          axis.title = element_text(face="bold",size = 9),
          strip.text = element_text(face="bold",size = 9), 
          panel.spacing.x = unit(0.6,units = "cm"), 
          axis.title.y = element_blank(), 
          legend.position = "none", 
          plot.title = element_text(hjust = 0.5))
  if(!is.null(xlim)){
    p <- p + coord_cartesian(xlim = xlim)
  }
  print(p)
}


## combine torus enrichment results across traits
combine_torus_traits <- function(trait_name_list, dir_results){
  registerDoParallel(cores = 10)
  
  result_torus <- foreach(trait = trait_name_list, .combine = rbind)%dopar%{
    torus_results <- read.table(paste0(dir_results,"/", trait, "/", trait, "_", "brain.results"), header = T, stringsAsFactors = F)
    torus.brain <- torus_results[torus_results$term != "Intercept",]
    
    torus_results <- read.table(paste0(dir_results,"/", trait, "/", trait, "_", "heart.results"), header = T, stringsAsFactors = F)
    torus.heart <- torus_results[torus_results$term != "Intercept",]
    
    torus_results <- read.table(paste0(dir_results,"/", trait, "/", trait, "_", "kidney.results"), header = T, stringsAsFactors = F)
    torus.kidney <- torus_results[torus_results$term != "Intercept",]
    
    torus_results <- read.table(paste0(dir_results,"/", trait, "/", trait, "_", "liver.results"), header = T, stringsAsFactors = F)
    torus.liver <- torus_results[torus_results$term != "Intercept",]
    
    torus_results <- read.table(paste0(dir_results,"/", trait, "/", trait, "_", "lung.results"), header = T, stringsAsFactors = F)
    torus.lung <- torus_results[torus_results$term != "Intercept",]
    
    torus_results <- read.table(paste0(dir_results,"/", trait, "/", trait, "_", "muscle.results"), header = T, stringsAsFactors = F)
    torus.muscle <- torus_results[torus_results$term != "Intercept",]
    
    torus_results <- read.table(paste0(dir_results,"/", trait, "/", trait, "_", "placenta.results"), header = T, stringsAsFactors = F)
    torus.placenta <- torus_results[torus_results$term != "Intercept",]
    
    torus_results <- read.table(paste0(dir_results,"/", trait, "/", trait, "_", "stomach.results"), header = T, stringsAsFactors = F)
    torus.stomach <- torus_results[torus_results$term != "Intercept",]
    
    torus.combined <- rbind(torus.brain, torus.heart, torus.kidney, torus.liver, torus.lung, torus.muscle, torus.placenta, torus.stomach)
    torus.combined <- cbind(Disease = trait, torus.combined)
    torus.combined
  }
  return(result_torus)
}

## change names for traits
change_annot_names <- function(annot_list){
  annot_list <- gsub("^brain_hg19_mod.bed.1$","brain", annot_list)
  annot_list <- gsub("^heart_hg19_mod.bed.1$","heart", annot_list)
  annot_list <- gsub("^kidney_hg19_mod.bed.1$","kidney", annot_list)
  annot_list <- gsub("^liver_hg19_mod.bed.1$","liver", annot_list)
  annot_list <- gsub("^lung_hg19_mod.bed.1$","lung", annot_list)
  annot_list <- gsub("^muscle_hg19_mod.bed.1$","muscle", annot_list)
  annot_list <- gsub("^placenta_hg19_mod.bed.1$","placenta", annot_list)
  annot_list <- gsub("^stomach_hg19_mod.bed.1$","stomach", annot_list)
  return(annot_list)
}

#convert from ln units
convert_values <- function(column){
  return(exp(column))
}

```


```{r torus data import}
dir_results <- "/project2/xinhe/cdu/torus/results/Xiao"
result_torus <- combine_torus_traits(trait_name_list, dir_results)

result_torus$term <- change_annot_names(result_torus$term)
names(result_torus)[names(result_torus) == "term"] <- "Category"
names(result_torus)[names(result_torus) == "estimate"] <- "Enrichment"
names(result_torus)[names(result_torus) == "low"] <- "Low"
names(result_torus)[names(result_torus) == "high"] <- "High"

result_torus <- cbind(result_torus[1:2], apply(result_torus[3:5],2, convert_values) )

result_torus$Category <- factor(result_torus$Category, levels = rev(annot_list) )
result_torus$Color <- factor(result_torus$Category, levels = annot_list)
```


```{r torus data table}
DT::datatable(format(result_torus[,1:5], digits = 2), 
              class = 'cell-border stripe',
              style = 'bootstrap',
              options = list(scrollX = TRUE, keys = TRUE, pageLength = length(annot_list)),
              rownames = F)
```


```{r torus plots, fig.height = 9}
ggplot_enrichment(result_torus, title = "", xlim = c(0,50))
ggplot_log2_enrichment(result_torus, title = "", xlim=c(0,6))
```


